File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/LinkedList/Pseudocodes/Addanddelete.py
# Add and Delete in Linked List - Inventory (MotorPH)


class Node:
    def __init__(self, brand, model, quantity):
        self.data = {"brand": brand, "model": model, "quantity": quantity}
        self.next = None


def add_at_head(head, new_node):
    """Add node at beginning."""
    new_node.next = head
    return new_node


def delete_by_brand(head, target_brand):
    """Delete first node with matching brand. Returns (new_head, deleted)."""
    if head is None:
        return None, False
    if head.data["brand"] == target_brand:
        return head.next, True
    current = head
    previous = None
    while current:
        if current.data["brand"] == target_brand:
            previous.next = current.next
            return head, True
        previous = current
        current = current.next
    return head, False


def display(head):
    """Display linked list from head."""
    if head is None:
        print("  (empty)")
        return
    current = head
    while current:
        d = current.data
        print(f"  - {d['brand']} {d['model']}: {d['quantity']} units")
        current = current.next


def main():
    head = None

    # Add nodes at beginning
    head = add_at_head(head, Node("Suzuki", "Burgman Street", 7))
    head = add_at_head(head, Node("Yamaha", "NMAX 155", 5))
    head = add_at_head(head, Node("Honda", "Click 125", 10))

    print("Linked list after adds (at head):")
    display(head)

    # Delete node by brand
    head, deleted = delete_by_brand(head, "Yamaha")
    print(f"\nDelete 'Yamaha': {'deleted' if deleted else 'not found'}")
    print("Linked list after delete:")
    display(head)

    # Delete another
    head, deleted = delete_by_brand(head, "Kawasaki")
    print(f"\nDelete 'Kawasaki': {'deleted' if deleted else 'not found'}")
    print("Linked list after delete:")
    display(head)


if __name__ == "__main__":
    main()

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/LinkedList/Pseudocodes/Addanddelete.java
package LinkedList.Pseudocodes;

/**
 * Add and Delete in Linked List - Inventory (MotorPH)
 */
public class Addanddelete {

    static class Node {
        InventoryItem data;
        Node next;

        Node(String brand, String model, int quantity) {
            this.data = new InventoryItem(brand, model, quantity);
            this.next = null;
        }
    }

    record InventoryItem(String brand, String model, int quantity) {}

    /** Add node at beginning. Returns new head. */
    static Node addAtHead(Node head, Node newNode) {
        newNode.next = head;
        return newNode;
    }

    record DeleteResult(Node head, boolean deleted) {}

    /** Delete first node with matching brand. Returns new head and whether a node was deleted. */
    static DeleteResult deleteByBrand(Node head, String targetBrand) {
        if (head == null) {
            return new DeleteResult(null, false);
        }
        if (head.data.brand().equals(targetBrand)) {
            return new DeleteResult(head.next, true);
        }
        Node current = head;
        Node previous = null;
        while (current != null) {
            if (current.data.brand().equals(targetBrand)) {
                previous.next = current.next;
                return new DeleteResult(head, true);
            }
            previous = current;
            current = current.next;
        }
        return new DeleteResult(head, false);
    }

    /** Display linked list from head. */
    static void display(Node head) {
        if (head == null) {
            System.out.println("  (empty)");
            return;
        }
        Node current = head;
        while (current != null) {
            InventoryItem d = current.data;
            System.out.println("  - " + d.brand() + " " + d.model() + ": " + d.quantity() + " units");
            current = current.next;
        }
    }

    public static void main(String[] args) {
        Node head = null;

        // Add nodes at beginning
        head = addAtHead(head, new Node("Suzuki", "Burgman Street", 7));
        head = addAtHead(head, new Node("Yamaha", "NMAX 155", 5));
        head = addAtHead(head, new Node("Honda", "Click 125", 10));

        System.out.println("Linked list after adds (at head):");
        display(head);

        // Delete node by brand
        DeleteResult r = deleteByBrand(head, "Yamaha");
        head = r.head();
        System.out.println("\nDelete 'Yamaha': " + (r.deleted() ? "deleted" : "not found"));
        System.out.println("Linked list after delete:");
        display(head);

        // Delete another (not present)
        r = deleteByBrand(head, "Kawasaki");
        head = r.head();
        System.out.println("\nDelete 'Kawasaki': " + (r.deleted() ? "deleted" : "not found"));
        System.out.println("Linked list after delete:");
        display(head);
    }
}
